Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp

```diff
@@ -848,55 +1176,117 @@ class Gx2Fitter {
         if (!stateHasMeasurement && !doMaterial) {
           ACTS_INFO("    Skip state.");
           continue;
         }
 
-        jacobianFromStart = trackState.jacobian() * jacobianFromStart;
+        // update all Jacobians from start
+        for (auto& jac : jacobianFromStart) {
+          jac = trackState.jacobian() * jac;
+        }
 
         // Handle measurement
         if (stateHasMeasurement) {
-          ACTS_VERBOSE("    Handle measurement");
+          ACTS_DEBUG("    Handle measurement.");
 
           auto measDim = trackState.calibratedSize();
           countNdf += measDim;
 
           if (measDim == 1) {
-            addToGx2fSums<1>(aMatrix, bVector, chi2sum, jacobianFromStart,
-                             trackState, *m_addToSumLogger);
+            addMeasurementToGx2fSums<1>(aMatrixExtended, bVectorExtended,
+                                        chi2sum, jacobianFromStart, trackState,
+                                        *m_addToSumLogger);
           } else if (measDim == 2) {
-            addToGx2fSums<2>(aMatrix, bVector, chi2sum, jacobianFromStart,
-                             trackState, *m_addToSumLogger);
+            addMeasurementToGx2fSums<2>(aMatrixExtended, bVectorExtended,
+                                        chi2sum, jacobianFromStart, trackState,
+                                        *m_addToSumLogger);
           } else if (measDim == 3) {
-            addToGx2fSums<3>(aMatrix, bVector, chi2sum, jacobianFromStart,
-                             trackState, *m_addToSumLogger);
+            addMeasurementToGx2fSums<3>(aMatrixExtended, bVectorExtended,
+                                        chi2sum, jacobianFromStart, trackState,
+                                        *m_addToSumLogger);
           } else if (measDim == 4) {
-            addToGx2fSums<4>(aMatrix, bVector, chi2sum, jacobianFromStart,
-                             trackState, *m_addToSumLogger);
+            addMeasurementToGx2fSums<4>(aMatrixExtended, bVectorExtended,
+                                        chi2sum, jacobianFromStart, trackState,
+                                        *m_addToSumLogger);
           } else if (measDim == 5) {
-            addToGx2fSums<5>(aMatrix, bVector, chi2sum, jacobianFromStart,
-                             trackState, *m_addToSumLogger);
+            addMeasurementToGx2fSums<5>(aMatrixExtended, bVectorExtended,
+                                        chi2sum, jacobianFromStart, trackState,
+                                        *m_addToSumLogger);
           } else if (measDim == 6) {
-            addToGx2fSums<6>(aMatrix, bVector, chi2sum, jacobianFromStart,
-                             trackState, *m_addToSumLogger);
+            addMeasurementToGx2fSums<6>(aMatrixExtended, bVectorExtended,
+                                        chi2sum, jacobianFromStart, trackState,
+                                        *m_addToSumLogger);
```

Comment:
I think this could be simplified with `visit_measurement`

Reply:
Probably a bit cleaner :)